cpu/kinetis/periph/i2c.c

```diff
@@ -45,474 +51,502 @@
+            DEBUG("i2c: start timeout\n");
+            return -EAGAIN;
+        }
+        /* slave -> master transition triggers the initial start condition */
+        i2c->C1 |= I2C_C1_IICEN_MASK | I2C_C1_MST_MASK | I2C_C1_TX_MASK;
     }
-
-    return 0;
-}
-
-static inline int _i2c_receive(I2C_Type *dev, uint8_t *data, int length)
-{
-    int n = 0;
-
-    /* set receive mode */
-    dev->C1 = I2C_C1_IICEN_MASK | I2C_C1_MST_MASK;
-
-    if (length == 1) {
-        /* Send NACK after receiving the next byte */
-        dev->C1 |= I2C_C1_TXAK_MASK;
+    i2c_state[dev].active = 1;
+    TRACE("i2c: Start C1=%02x S=%02x\n", (unsigned)i2c->C1, (unsigned)i2c->S);
+    if (i2c->S & I2C_S_ARBL_MASK) {
+        DEBUG("i2c: arbitration lost\n");
+        i2c_clear_irq_flags(i2c);
+        bit_clear8(&i2c->C1, I2C_C1_MST_SHIFT);
+        i2c_state[dev].active = 0;
+        return -EAGAIN;
     }
-
-    /* Initiate master receive mode by reading the data register */
-    dev->D;
-
-    while (length > 0) {
-        while (!(dev->S & I2C_S_IICIF_MASK));
-
-        dev->S = I2C_S_IICIF_MASK;
-
-        if (_i2c_arbitration_lost(dev)) {
-            DEBUG("i2c:_receive: arbitration lost (to go=%d)\n", length);
-            return -1;
+    if (flags & I2C_ADDR10) {
+        /* 10 bit addressing */
+        /* Send 10 bit address tag + 2 msb + R/W flag */
+        uint8_t msb = (addr >> 8);
+        uint8_t lsb = (addr & 0xff);
+        uint8_t addr_rw = I2C_10BIT_MAGIC | (msb << 1);
+        if (read_flag) {
+            addr_rw |= I2C_READ;
         }
-
-        length--;
-
-        if (length == 1) {
-            /* Send NACK after receiving the next byte */
-            dev->C1 |= I2C_C1_TXAK_MASK;
+        int res = i2c_tx_addr(dev, addr_rw);
+        if (res < 0) {
+            return res;
         }
-
-        if (length == 0) {
-            /* Stop immediately because the receiving of the next byte will be
-             * initiated by reading the data register (dev->D). */
-            dev->C1 &= ~I2C_C1_MST_MASK;
+        if (!read_flag) {
+            /* Send lower 8 bits of address */
+            res = i2c_tx_addr(dev, lsb);
         }
-
-        data[n] = dev->D;
-        TRACE("i2c: rx: %02x\n", (unsigned int)data[n]);
-        n++;
+        return res;
     }
-
-    return n;
+    /* 7 bit addressing */
+    /* Send address + R/W flag */
+    uint8_t addr_rw = (addr << 1);
+    if (read_flag) {
+        addr_rw |= I2C_READ;
+    }
+    return i2c_tx_addr(dev, addr_rw);
 }
 
-static inline int _i2c_transmit(I2C_Type *dev, const uint8_t *data, int length)
+int i2c_read_bytes(i2c_t dev, uint16_t addr, void *data, size_t len, uint8_t flags)
 {
-    int n = 0;
-
-    while (length > 0) {
-        TRACE("i2c: tx: %02x\n", data[n]);
-        dev->D = data[n];
-
-        while (!(dev->S & I2C_S_IICIF_MASK));
-
-        dev->S = I2C_S_IICIF_MASK;
+    I2C_Type *i2c = i2c_config[dev].i2c;
```

Comment:
Could you add a guard here please ? 
```
if (len == 0 || data == NULL) {
    return -EINVAL;
}
```

Reply:
Could you also return in case of unsupported features ?

Reply:
`len == 0` is not an error, it can be used to send a start condition or stop condition, depending on flags.
I don't think there are any unsupported flags here (other than REG16, which I think we can just ignore in the _bytes functions)

Reply:
> it can be used to send a start condition or stop condition, depending on flags

I never tought of it but is this really useful for the user ? I mean, we're suppose to read or write on the bus so why just sending a start out of the blue ?

Reply:
The hih6130 is one weird device which uses empty writes to trigger measurements. The user sends a start condition for writing followed by a stop condition, so zero data bytes. Then when reading the device will return the most recent measurement.

Reply:
> The hih6130 is one weird device which uses empty writes to trigger measurements

Indeed, I'll look at this one. Is this even I2C/TWI compliant ?

Reply:
Dunno, but it's a device that we have a driver for